File: backend/src/main/java/com/hostel/backend/service/AuthService.java
package com.hostel.backend.service;

import com.hostel.backend.entity.Block;
import com.hostel.backend.entity.Hostel;
import com.hostel.backend.entity.User;
import com.hostel.backend.enums.Role;
import com.hostel.backend.exception.AppException;
import com.hostel.backend.repository.BlockRepository;
import com.hostel.backend.repository.HostelRepository;
import com.hostel.backend.repository.UserRepository;
import com.hostel.backend.dto.RegisterRequest;
import com.hostel.backend.security.JwtUtil;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Map;

@Service
public class AuthService {

    private final UserRepository    userRepository;
    private final PasswordEncoder   passwordEncoder;
    private final JwtUtil           jwtUtil;
    private final OtpService        otpService;
    private final HostelRepository  hostelRepository;
    private final BlockRepository   blockRepository;

    public AuthService(UserRepository userRepository,
                       PasswordEncoder passwordEncoder,
                       JwtUtil jwtUtil,
                       OtpService otpService,
                       HostelRepository hostelRepository,
                       BlockRepository blockRepository) {
        this.userRepository   = userRepository;
        this.passwordEncoder  = passwordEncoder;
        this.jwtUtil          = jwtUtil;
        this.otpService       = otpService;
        this.hostelRepository = hostelRepository;
        this.blockRepository  = blockRepository;
    }

    @Transactional
    public User register(RegisterRequest request) {

        // ── Role guard ─────────────────────────────────────────────────────
        if (request.getRole() != null && request.getRole() != Role.STUDENT) {
            throw new AppException("Self-registration is only allowed for STUDENT role", HttpStatus.FORBIDDEN);
        }

        // ── Required field validation ──────────────────────────────────────
        if (request.getName() == null || request.getEmail() == null ||
            request.getPassword() == null || request.getPhoneNumber() == null) {
            throw new AppException("Required fields are missing", HttpStatus.BAD_REQUEST);
        }

        // ── College email validation ───────────────────────────────────────
        if (!request.getEmail().endsWith("@stu.manit.ac.in")) {
            throw new AppException("Students must register with college email (@stu.manit.ac.in)", HttpStatus.BAD_REQUEST);
        }

        // ── Student-specific required fields ──────────────────────────────
        if (request.getScholarNumber() == null || request.getRoomNumber() == null) {
            throw new AppException("scholarNumber and roomNumber are required", HttpStatus.BAD_REQUEST);
        }
        if (request.getHostelId() == null || request.getBlockId() == null) {
            throw new AppException("hostelId and blockId are required", HttpStatus.BAD_REQUEST);
        }

        // ── Handle existing email ──────────────────────────────────────────
        userRepository.findByEmail(request.getEmail()).ifPresent(existing -> {
            if (existing.isActive()) {
                throw new AppException("Email already registered", HttpStatus.CONFLICT);
            }
            userRepository.delete(existing);
            userRepository.flush();
        });

        // ── Handle duplicate phone ─────────────────────────────────────────
        userRepository.findByPhoneNumber(request.getPhoneNumber()).ifPresent(existing -> {
            if (!existing.isActive()) {
                userRepository.delete(existing);
                userRepository.flush();
            } else {
                throw new AppException("Phone number already registered", HttpStatus.CONFLICT);
            }
        });

        // ── Hostel + Block lookup ──────────────────────────────────────────
        Hostel hostel = hostelRepository.findById(request.getHostelId())
                .orElseThrow(() -> new AppException("Hostel not found", HttpStatus.BAD_REQUEST));
        Block block = blockRepository.findById(request.getBlockId())
                .orElseThrow(() -> new AppException("Block not found", HttpStatus.BAD_REQUEST));

        // ── Block must belong to the selected hostel ───────────────────────
        if (!block.getHostel().getId().equals(hostel.getId())) {
            throw new AppException("Block does not belong to the selected hostel", HttpStatus.BAD_REQUEST);
        }

        // ✅ Room capacity check — max 2 active students per room in a block ─
        long occupants = userRepository.countByBlockIdAndRoomNumberAndActiveTrue(
                block.getId(), request.getRoomNumber());
        if (occupants >= 2) {
            throw new AppException(
                "Room " + request.getRoomNumber() + " in this block already has 2 students. " +
                "Please contact the caretaker if you believe this is an error.",
                HttpStatus.CONFLICT
            );
        }

        // ── Send OTP first — before saving user ───────────────────────────
        try {
            otpService.generateAndSendOtp(request.getEmail());
        } catch (Exception e) {
            throw new AppException("Failed to send OTP email. Please try again.", HttpStatus.INTERNAL_SERVER_ERROR);
        }

        // ── Save user only after OTP sent successfully ─────────────────────
        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(Role.STUDENT)
                .phoneNumber(request.getPhoneNumber())
                .scholarNumber(request.getScholarNumber())
                .roomNumber(request.getRoomNumber())
                .hostel(hostel)
                .block(block)
                .active(false)
                .build();

        return userRepository.save(user);
    }

    public Map<String, String> login(String email, String password) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException("User not found", HttpStatus.NOT_FOUND));

        if (!user.isActive()) {
            throw new AppException("Account not verified. Please verify your email first.", HttpStatus.FORBIDDEN);
        }
        if (!passwordEncoder.matches(password, user.getPassword())) {
            throw new AppException("Invalid credentials", HttpStatus.UNAUTHORIZED);
        }

        String token = jwtUtil.generateToken(user.getEmail(), user.getRole().name());
        return Map.of(
            "token",  token,
            "role",   user.getRole().name(),
            "name",   user.getName(),
            "email",  user.getEmail()
        );
    }

    // ✅ FIXED — Forgot password now sends OTP via email correctly
    @Transactional
    public void forgotPassword(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException("No account found with this email", HttpStatus.NOT_FOUND));

        if (!user.isActive()) {
            throw new AppException("Account is not verified. Please complete email verification first.", HttpStatus.FORBIDDEN);
        }

        try {
            otpService.generateAndSendOtp(email);
        } catch (Exception e) {
            throw new AppException("Failed to send OTP email. Please try again.", HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }

    // ✅ FIXED — Reset password verifies the OTP then updates the password
    @Transactional
    public void resetPassword(String email, String otp, String newPassword) {
        // Verify OTP — throws AppException if invalid/expired
        otpService.verifyOtpForPasswordReset(email, otp);

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AppException("User not found", HttpStatus.NOT_FOUND));

        if (newPassword == null || newPassword.trim().length() < 6) {
            throw new AppException("Password must be at least 6 characters", HttpStatus.BAD_REQUEST);
        }

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }
}

File: backend/src/services/auth.service.js
const prisma = require('../config/prisma');
const { AppError } = require('../utils/AppError');
const { hashPassword, comparePassword } = require('../utils/password');
const { generateToken } = require('../utils/jwt');
const otpService = require('./otp.service');

const VALID_BLOCKS = ['A', 'B'];

// Shared by the public self-registration flow (`register`, below — OTP
// required, starts inactive, matching the original Java behavior exactly)
// and admin-created student accounts (`admin.service.js#createStudent` —
// no OTP, active immediately, since an admin creating the account already
// vouches for it, the same way admin-created warden/caretaker/worker
// accounts always have been). `sendOtp: false` skips the email step and
// activates the account right away instead.
async function createStudentAccount(request, { sendOtp = true } = {}) {
  const { name, email, password, phoneNumber, scholarNumber, roomNumber, block } = request;

  if (!name || !email || !password || !phoneNumber) {
    throw new AppError('Required fields are missing', 400);
  }
  if (!email.endsWith('@stu.manit.ac.in')) {
    throw new AppError('Students must register with college email (@stu.manit.ac.in)', 400);
  }
  if (!scholarNumber || !roomNumber) {
    throw new AppError('scholarNumber and roomNumber are required', 400);
  }
  if (!block || !VALID_BLOCKS.includes(block)) {
    throw new AppError('block is required and must be A or B', 400);
  }

  const existingByEmail = await prisma.user.findUnique({ where: { email } });
  if (existingByEmail) {
    if (existingByEmail.active) throw new AppError('Email already registered', 409);
    await prisma.user.delete({ where: { id: existingByEmail.id } });
  }

  const existingByPhone = await prisma.user.findUnique({ where: { phoneNumber } });
  if (existingByPhone) {
    if (!existingByPhone.active) {
      await prisma.user.delete({ where: { id: existingByPhone.id } });
    } else {
      throw new AppError('Phone number already registered', 409);
    }
  }

  const occupants = await prisma.user.count({
    where: { block, roomNumber, active: true },
  });
  if (occupants >= 2) {
    throw new AppError(
      `Room ${roomNumber} in this block already has 2 students. Please contact the caretaker if you believe this is an error.`,
      409
    );
  }

  // Send OTP first — before saving the user, same order as the original.
  if (sendOtp) {
    await otpService.generateAndSendOtp(email);
  }

  const user = await prisma.user.create({
    data: {
      name,
      email,
      password: await hashPassword(password),
      role: 'STUDENT',
      phoneNumber,
      scholarNumber,
      roomNumber,
      block,
      active: !sendOtp,
      appEnabled: true,
    },
  });

  return user;
}

// Mirrors AuthService.java `register()`, minus the hostel/block foreign-key
// lookups — this app now serves exactly one hostel (Hostel 10), so `block`
// is a plain 'A'|'B' enum value instead of a hostelId+blockId pair that had
// to be resolved and cross-checked against a Hostel/Block table. See
// docs/MIGRATION_NOTES.md "Hostel-10 scoping" for why those tables are gone.
async function register(request) {
  if (request.role && request.role !== 'STUDENT') {
    throw new AppError('Self-registration is only allowed for STUDENT role', 403);
  }
  return createStudentAccount(request, { sendOtp: true });
}

async function login(email, password) {
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) throw new AppError('User not found', 404);

  if (!user.active) {
    throw new AppError('Account not verified. Please verify your email first.', 403);
  }
  if (!(await comparePassword(password, user.password))) {
    throw new AppError('Invalid credentials', 401);
  }

  const token = generateToken(user.email, user.role);
  return { token, role: user.role, name: user.name, email: user.email };
}

// Mirrors AuthService.java's forgotPassword/resetPassword (OTP-based)
// exactly. Unlike passwordReset.service.js's token-based flow, no frontend
// page ever called this variant — the "Forgot password?" link uses the
// token-based flow instead (see auth.controller.js). Kept unwired/unused,
// same as the original Spring Boot app, since removing it isn't necessary
// and two parallel reset flows already existed pre-migration.
async function forgotPassword(email) {
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) throw new AppError('No account found with this email', 404);
  if (!user.active) {
    throw new AppError('Account is not verified. Please complete email verification first.', 403);
  }
  await otpService.generateAndSendOtp(email);
}

async function resetPassword(email, otp, newPassword) {
  await otpService.verifyOtpForPasswordReset(email, otp);

  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) throw new AppError('User not found', 404);

  if (!newPassword || newPassword.trim().length < 6) {
    throw new AppError('Password must be at least 6 characters', 400);
  }

  await prisma.user.update({
    where: { id: user.id },
    data: { password: await hashPassword(newPassword) },
  });
}

module.exports = { register, login, forgotPassword, resetPassword, createStudentAccount };
